packages/pds/src/db/message-queue/index.ts

```diff
@@ -0,0 +1,316 @@
+import * as schema from '../../lexicon/schemas'
+import { TID } from '@atproto/common'
+import { MessageQueue } from '../types'
+
+type GetAuthStoreFn = (did: string) => AuthStore
+
+export class SqlMessageQueue implements MessageQueue {
+  private cursorExists = false
+
+  constructor(
+    private name: string,
+    private db: Database,
+    private getAuthStore: GetAuthStoreFn,
+  ) {}
+
+  async send(tx: Database, messages: Message | Message[]): Promise<void> {
+    const msgArray = Array.isArray(messages) ? messages : [messages]
+    if (msgArray.length === 0) return
+    const now = new Date().toISOString()
+    const values = msgArray.map((msg) => ({
+      message: JSON.stringify(msg),
+      createdAt: now,
+    }))
+
+    await tx.db.insertInto('message_queue').values(values).execute()
+
+    this.processNext()
+  }
+
+  private async ensureCursor(): Promise<void> {
+    if (this.cursorExists) return
+    await this.db.db
+      .insertInto('message_queue_cursor')
+      .values({ consumer: this.name, cursor: 1 })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    this.cursorExists = true
+  }
+
+  async processAll(): Promise<void> {
+    await this.ensureCursor()
+    await this.db.transaction(async (dbTxn) => {
+      let builder = dbTxn.db
+        .selectFrom('message_queue_cursor as cursor')
+        .innerJoin('message_queue', (join) =>
+          join.onRef('cursor.cursor', '<=', 'message_queue.id'),
+        )
+        .where('cursor.consumer', '=', this.name)
+        .selectAll()
+      if (this.db.dialect !== 'sqlite') {
+        builder = builder.forUpdate()
+      }
+      const res = await builder.execute()
+      // all caught up
+      if (res.length === 0) return
+
+      for (const row of res) {
+        const message: Message = JSON.parse(row.message)
+        await this.handleMessage(dbTxn, message)
+      }
+      const nextCursor = res[res.length - 1].id + 1
+      await dbTxn.db
+        .updateTable('message_queue_cursor')
+        .set({ cursor: nextCursor })
+        .where('consumer', '=', this.name)
+        .execute()
+    })
+  }
+
+  async processNext(): Promise<void> {
+    await this.ensureCursor()
+    await this.db.transaction(async (dbTxn) => {
+      let builder = dbTxn.db
+        .selectFrom('message_queue_cursor as cursor')
+        .innerJoin('message_queue', (join) =>
+          join.onRef('cursor.cursor', '<=', 'message_queue.id'),
+        )
+        .where('cursor.consumer', '=', this.name)
+        .orderBy('id', 'asc')
+        .limit(1)
+        .selectAll()
+      if (this.db.dialect !== 'sqlite') {
+        builder = builder.forUpdate()
+      }
+
+      const res = await builder.executeTakeFirst()
+      // all caught up
+      if (!res) return
+
+      const message: Message = JSON.parse(res.message)
+      await this.handleMessage(dbTxn, message)
+      await dbTxn.db
+        .updateTable('message_queue_cursor')
+        .set({ cursor: sql`cursor + 1` })
```

Comment:
I think by incrementing by one here we can cause messages to get processed multiple times if there's a gap in the ids.  Pretty sure we'll want to catch the cursor up to whichever message was just processed (plus 1).

Reply:
a true hero you are 🙏